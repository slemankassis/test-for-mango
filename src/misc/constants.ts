export const CURRENCY_SYMBOL = 'â‚¬';
export const UNKNOWN_ERROR = 'Unknown error';
