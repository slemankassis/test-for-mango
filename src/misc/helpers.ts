import { RangeValue } from './models/RangeValue';

export function generatePercentages(arr: number[]): RangeValue[] {
  const sortedArr = [...arr].sort((a, b) => a - b);

  const mul = 100 / (sortedArr[sortedArr.length - 1] - sortedArr[0]);
  return sortedArr.map((value) => {
    return {
      value,
      percent: (value - sortedArr[0]) * mul,
    };
  });
}

export function closestRangeValueByPercentage(
  percentage: number,
  rangeValues: RangeValue[],
): RangeValue {
  return rangeValues.reduce((prev, curr) => {
    return Math.abs(curr.percent - percentage) < Math.abs(prev.percent - percentage)
      ? curr
      : prev;
  });
}

export function closestRangeValueByValue(
  value: number,
  rangeValues: RangeValue[],
): RangeValue {
  return rangeValues.reduce((prev, curr) => {
    return Math.abs(curr.value - value) < Math.abs(prev.value - value) ? curr : prev;
  });
}
