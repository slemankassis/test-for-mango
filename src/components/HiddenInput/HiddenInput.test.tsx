import React from 'react';
import { cleanup, render } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import HiddenInput from './HiddenInput';
import userEvent from '@testing-library/user-event';
import { CURRENCY_SYMBOL } from '../../misc/constants';

afterEach(cleanup);

const setup = () => {
  const utils = render(
    <HiddenInput value={100} unit={CURRENCY_SYMBOL} ariaLabel="test input" />,
  );
  const input = utils.getByLabelText('test input');
  return {
    input,
    ...utils,
  };
};

describe('HiddenInput operations', () => {
  test('should show the value', () => {
    const { getByDisplayValue } = setup();

    expect(getByDisplayValue(100)).toHaveValue(100);
  });

  test('should show the correct unit', () => {
    const { getByText } = setup();

    expect(getByText(/€/i).textContent).toBe('€');
  });

  test('should be disabled', () => {
    const { getByLabelText } = render(
      <HiddenInput value={100} unit="€" ariaLabel="test input" disabled={true} />,
    );

    expect(getByLabelText('test input')).toBeDisabled();
  });

  test('should allow numbers', () => {
    const { input } = setup();

    userEvent.clear(input);
    userEvent.type(input, '12');

    expect(input).toHaveValue(12);
  });

  test('should not allow non numeric characters', () => {
    const { input } = setup();

    userEvent.clear(input);
    userEvent.type(input, 'abc');

    expect(input).toHaveValue(null);
  });

  test('should show a new value comming from props', () => {
    const { input, rerender } = setup();

    userEvent.clear(input);
    userEvent.type(input, '12');

    rerender(<HiddenInput value={999} ariaLabel="test input" />);

    expect(input).toHaveValue(999);
  });

  test('should not modify user input when the same value comming from props', () => {
    const { input, rerender } = setup();

    userEvent.clear(input);
    userEvent.type(input, '12');

    rerender(<HiddenInput value={100} ariaLabel="test input" />);

    expect(input).toHaveValue(12);
  });

  test('should user can clear the input', () => {
    const { input } = setup();

    userEvent.clear(input);

    expect(input).toHaveValue(null);
  });
});

describe('HiddenInput snapshots', () => {
  test('snapshot match', () => {
    const tree = TestRenderer.create(<HiddenInput value={100} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
