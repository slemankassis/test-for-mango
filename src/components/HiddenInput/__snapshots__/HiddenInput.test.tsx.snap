// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HiddenInput snapshots snapshot match 1`] = `
<div
  className="sc-bdfBQB jrfRxd"
>
  <input
    className="span-input"
    onBlur={[Function]}
    onChange={[Function]}
    style={
      Object {
        "width": "3ch",
      }
    }
    type="number"
    value="100"
  />
  <span />
</div>
`;
